x/payment/types/wirepayfordata.go

```diff
@@ -221,3 +215,33 @@ func ProcessWirePayForData(msg *MsgWirePayForData, squareSize uint64) (*tmproto.
 
 	return &coreMsg, pfd, shareCommit.Signature, nil
 }
+
+func HasWirePayForData(tx sdk.Tx) bool {
+	for _, msg := range tx.GetMsgs() {
+		msgName := sdk.MsgTypeURL(msg)
+		if msgName == URLMsgWirePayForData {
+			return true
+		}
+	}
+	return false
+}
+
+func ExtractMsgWirePayForData(tx sdk.Tx) (*MsgWirePayForData, error) {
+	noWirePFDError := errors.New("sdk.Tx does not contain MsgWirePayForData sdk.Msg")
+	if !HasWirePayForData(tx) {
+		return nil, noWirePFDError
+	}
+
+	// only support malleated transactions that contain a single sdk.Msg
+	if len(tx.GetMsgs()) != 1 {
+		return nil, errors.New("sdk.Txs with a single MsgWirePayForData are currently supported")
+	}
+
+	msg := tx.GetMsgs()[0]
+	wireMsg, ok := msg.(*MsgWirePayForData)
+	if !ok {
+		return nil, noWirePFDError
```

Comment:
This looks like it should be a panic, or a different error. The calling function of this is only prepared to handle the first error on line 232.

Reply:
> The calling function of this is only prepared to handle the first error on line 232.

I think I'm missing something here, why is the calling function only equipped to handle the first error but not this one?

Reply:
This function is only used when preparing a proposal, which assumes that transactions have either 0 or 1 data. If it has 0, then it can be included in a block via some other means.

The issue is that this function also handles cases where there could be more than 1 data. Such transactions will never be included in blocks and continue to congest the mempool forever. If they have high priority fees (whenever that lands in the Cosmos SDK), then they'll never be evicted but also never included. This applies more to the above error from checking `if len(tx.GetMsgs()) != 1`.

As for this error, shouldn't it be impossible given the precondition that `HasWirePayForData(tx)` is true?

Reply:
> This function is only used when preparing a proposal, which assumes that transactions have either 0 or 1 data. If it has 0, then it can be included in a block via some other means.

it's only used during estimating the number of shares that a message takes up and when verifying signatures of PfD during tests, not when determining if a tx is included in the block or not.

> As for this error, shouldn't it be impossible given the precondition that HasWirePayForData(tx) is true?

the precondition only quickly checks that the proto message's URL is `/payment.WirePayForData` which could be faked or from some old version of wirePfD. Something else could go wrong as well and we don't want to panic so we have to handle/bubble this error

Reply:
https://github.com/celestiaorg/celestia-app/blob/823063062789c6a2786ce1d06ea758d4c46bbd28/app/prepare_proposal.go#L97-L103

even with these comments, I can see this being confusing. This might be a premature abstraction that trades too much comprehension of the code for simply not copying the code in two places. I'm happy to revert it if we think that's a good idea

Reply:
Reverting probably isn't necessary, but could you take a stab at maybe improving the comments?

Reply:
yeah, definitely 37beabb42599d972394fd937697decf1f3d28737